src/shared_library.c

```diff
@@ -65,33 +74,116 @@ rcutils_load_shared_library(
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib, RCUTILS_RET_INVALID_ARGUMENT);
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(library_path, RCUTILS_RET_INVALID_ARGUMENT);
   RCUTILS_CHECK_ALLOCATOR(&allocator, return RCUTILS_RET_INVALID_ARGUMENT);
-
-  if (lib->library_path != NULL) {
-    lib->allocator.deallocate(lib->library_path, lib->allocator.state);
+  if (NULL != lib->lib_pointer) {
+    RCUTILS_SET_ERROR_MSG("lib argument is not zero-initialized");
+    return RCUTILS_RET_INVALID_ARGUMENT;
   }
 
+  rcutils_ret_t ret = RCUTILS_RET_OK;
   lib->allocator = allocator;
 
+#ifndef _WIN32
+  lib->lib_pointer = dlopen(library_path, RTLD_LAZY);
+  if (NULL == lib->lib_pointer) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("dlopen error: %s", dlerror());
+    return RCUTILS_RET_ERROR;
+  }
+
+#if defined(__APPLE__)
+  const char * image_name = NULL;
+  uint32_t image_count = _dyld_image_count();
+  for (uint32_t i = 0; NULL == image_name && i < image_count; ++i) {
```

Comment:
I feel like I'm missing something. Is my understanding correct that what ends up in `lib->library_path` is the absolute path of `library_path`? If so, can't we just call `realpath`? Or is it possible `library_path` and `lib->library_path` will point to two different files?

Reply:
> If so, can't we just call `realpath`?

The thing is, `library_path` might simply be the library name e.g. `libfoo.so`. And that doesn't mean it should be searched for in the current working directory. It's only after `dlopen` performs the search over RPATHs, (DY)LD_LIBRARY_PATHs, and RUNPATHs and actually finds it that you can get the full path to it.

Reply:
Ah, that makes sense. Thanks! I don't know what the comment etiquette is for this package, but I think it would be great to have that paragraph as a comment just to explain what the goal of this section of code is.

Reply:
Sure thing. See 1d4276a.